### ISSUE TITLE
Unit Multipliers

### ISSUE THREAD

**Issue by robons** [GSS-Cogs/csvwlib#71]

**Comment by rossbowen** [comment]
I think tonnes are a reasonable unit to make use of, but I can see the motivation for this more when thinking about currency, e.g., `pounds-sterling-thousands` and `pounds-sterling-millions` etc.

I think we spoke about abbreviated cubes recently - I just wanted to highlight this sentence:

https://www.w3.org/TR/vocab-data-cube/#dsd-dsd

> By default the values of all of the components will be attached to each individual observation, this is called the normalized representation. This allows such observations to stand alone, so that a SPARQL query to retrieve the observation can immediately locate the attributes which enable the observation to be interpreted. However, it is also permissible to attach attributes to the overall data set, to an intervening slice **_or to a specific Measure_** (in the case of multiple measures). This reduces some of the redundancy in the encoding of the instance data. To declare such an abbreviated structure, the qb:componentAttachment property of the specification should reference the class corresponding to the attachment level (e.g. qb:DataSet for attributes that will be attached to the overall data set). The classes which can be used as such attachment levels are all subclasses of **_qb:Attachable_**.

Interestingly a measure isn't a subclass of `qb:Attachable` but not sure if that's an accidental omission or not. To me it makes sense you'd like to be able to attach things like units to a measure.

Some thoughts:

- If attaching at the `qb:DataSet` level then things get complex when dealing with multiple measures,
- If attaching at the `qb:MeasureProperty` level,  if you attach to a measure which is defined elsewhere then you're introducing a statement which may not be true globally,
- At that point, if you create a bespoke measure in order to make a statement about units, I figure you may aswell just use measures which are a combo of `unit+multiplier`.

[`qudt`](http://qudt.org/2.1/vocab/unit) does have a distinct unit for `unit:MillionUSD` so doesn't feel like we're inconsistent with what's going on elsewhere.

```ttl
unit:MillionUSD
  a qudt:CurrencyUnit ;
  a qudt:DerivedUnit ;
  a qudt:Unit ;
  qudt:conversionMultiplier 1000000.0 ;
  qudt:hasDimensionVector qkdv:A0E0L0I0M0H0T0D1 ;
  qudt:hasQuantityKind quantitykind:Currency ;
  rdfs:isDefinedBy <http://qudt.org/2.1/vocab/unit> ;
  rdfs:label "Million US Dollars"@en ;
```

**Comment by ajtucker** [comment]
Agreed. We need to explore these component attachment details more and this is a good example.

I tend to think that data producers will balk at having to repeat information in extra columns and that data consumers will also be unnecessarily overwhelmed with these repeated values, so attaching these attribute values higher up in the cube a) gives us a way to encode this stuff without burdening the producer and b) gives PMD a hint that it can avoid displaying repeated values and instead roll them up into the column heading, for example.

Also, we want to move the responsibility for converting CSV to RDF downstream (to PMD), so the normalization can be done server side with e.g. SPARQL UPDATE as per the spec.

**Comment by ajtucker** [comment]
> Interestingly a measure isn't a subclass of `qb:Attachable` but not sure if that's an accidental omission or not. To me it makes sense you'd like to be able to attach things like units to a measure.

I reckon much of this comes from the SDMX model and think it's a bit cumbersome because of that.

As such we shouldn't constrain ourselves too much and could use this as an example of a more complex measure along the lines we've discussed before where the measured "value" of an observation isn't a literal, but a resource with an associated `rdf:type` so that we can assert properties of both the observation's value and also of the value's type.

**Comment by robons** [comment]
> If attaching at the qb:DataSet level then things get complex when dealing with multiple measures

I wouldn't attach the units/multipliers the qb:DataSet level if we've got multiple-measures - there's be no way to differentiate between which unit/multiplier went with each measure. 

If we did want to attach this information *higher up*, we could define a set of slices (one for each measure) and attach the units/multipliers at that level. 

> qudt does have a distinct unit for unit:MillionUSD so doesn't feel like we're inconsistent with what's going on elsewhere.

Yeah, I think that's a sensible approach given that they have a consistent definition mapping between the units - to show equivalence.

 > Also, we want to move the responsibility for converting CSV to RDF downstream (to PMD), so the normalization can be done server side with e.g. SPARQL UPDATE as per the spec.

I agree, but I think it'll take some time before we've agreed on the precise structure of the CSV-W's metadata that PMD will accept as input. So I'm working under the assumption that we'll have to do most, if not all of the transformation work ourselves to begin with in Jenkins.

> As such we shouldn't constrain ourselves too much and could use this as an example of a more complex measure along the lines we've discussed before where the measured "value" of an observation isn't a literal, but a resource with an associated rdf:type so that we can assert properties of both the observation's value and also of the value's type.

I think we can already uniquely associate an `sdmx-a:unitMult` with each distinct measure-observation given that we use qb's measure dimension approach.

----

I suppose a way forward with this problem would be to allow users to define new units which are derived from existing units with the `qudt:conversionMultiplier` specified. 

I agree it would be frustrating to have an additional column duplicating all of the data where most users will probably have consistent unitMultipliers for a given unit in a dataset. 

At least this way we have some notion of the units being related (and hence comparable) without having to specify the unitMult against every single observation.

We could have a separate issue looking more long-term at attaching this information higher up in the qb DSD? But I think this solution would allow us to move relatively quickly. 

Any thoughts on this idea?

**Comment by rossbowen** [comment]
I'm keen on giving users the ability to specify units using [qudt](https://github.com/qudt/qudt-public-repo/wiki/User-Guide-for-QUDT), though I wouldn't mandate that. Ideally we'll steer users to a well curated set of unit resources to pick from which would meet the most needs for most statistics. In the event the user has something a bit unique they need to specify, I wouldn't be asking them to do much more than asking them to provide a `rdfs:label` and `rdfs:comment`. Maybe we could type them as `qudt:Unit` but I need to read a bit more into `qudt` and see if there's a set of minimal requirements we need to meet for things to be valid.

I don't love the `sdmx-a:unitMult` approach - right now I don't think I'd be likely to advocate it to users. 

I think for now, a user needs to be able to specify custom units, but I wouldn't make them learn `qudt` or do too much work to be able to do that. Everything else I think is a nice to have but maybe not essential for the time being.

**Comment by robons** [comment]
Following discussion in a refinement session, we've decided to take the following action:

Allow the user to define something along the lines below:

`unit = NewQbUnitFromExisting(Thousands of turtles", "http://uri-of/existing/unit/turtles", 3)`

This should create a new unit which is related to the existing one using the qudt ontology to say that it's got a multiplying factory of 10^3 (in this case).
